index/scorch/snapshot_rollback.go

```diff
@@ -15,81 +15,142 @@
+		}
 
-			snapshot := snapshots.Bucket(k)
-			if snapshot == nil {
-				log.Printf("PreviousPersistedSnapshot:"+
-					" snapshot key, but bucket missing %x, continuing", k)
-				continue
-			}
+		snapshot := snapshots.Bucket(k)
+		if snapshot == nil {
+			log.Printf("RollbackPoints:"+
+				" snapshot key, but bucket missing %x, continuing", k)
+			continue
+		}
 
-			indexSnapshot, err := s.loadSnapshot(snapshot)
-			if err != nil {
-				log.Printf("PreviousPersistedSnapshot:"+
-					" unable to load snapshot, %v, continuing", err)
-				continue
+		meta := map[string][]byte{}
+		c2 := snapshot.Cursor()
+		for j, _ := c2.First(); j != nil; j, _ = c2.Next() {
+			if j[0] == boltInternalKey[0] {
+				internalBucket := snapshot.Bucket(j)
+				err = internalBucket.ForEach(func(key []byte, val []byte) error {
+					copiedVal := append([]byte(nil), val...)
+					meta[string(key)] = copiedVal
+					return nil
+				})
+				if err != nil {
+					break
+				}
 			}
+		}
 
-			indexSnapshot.epoch = snapshotEpoch
-			return indexSnapshot, nil
+		if err != nil {
+			log.Printf("RollbackPoints:"+
+				" failed in fetching interal data: %v", err)
+			continue
 		}
+
+		rollbackPoints = append(rollbackPoints, &RollbackPoint{
+			epoch: snapshotEpoch,
+			meta:  meta,
+		})
 	}
 
-	return nil, nil
+	return rollbackPoints, nil
 }
 
-// SnapshotRevert atomically brings the store back to the point in time
-// as represented by the revertTo IndexSnapshot. SnapshotRevert() should
-// only be passed an IndexSnapshot that came from the same store.
-func (s *Scorch) SnapshotRevert(revertTo *IndexSnapshot) error {
+// Rollback atomically and durably (if unsafeBatch is unset) brings
+// the store back to the point in time as represented by the
+// RollbackPoint. Rollbac() should only be passed a RollbackPoint
+// that came from the same store using the RollbackPoints() API.
+func (s *Scorch) Rollback(to *RollbackPoint) error {
+	if to == nil {
+		return fmt.Errorf("Rollback: RollbackPoint is nil")
+	}
+
+	if s.rootBolt == nil {
+		return fmt.Errorf("Rollback: root is nil")
+	}
+
+	// start a read-only bolt transaction
+	tx, err := s.rootBolt.Begin(false)
+	if err != nil {
+		return fmt.Errorf("Rollback: failed to start read-only transaction")
+	}
+
+	snapshots := tx.Bucket(boltSnapshotsBucket)
+	if snapshots == nil {
+		return fmt.Errorf("Rollback: no snapshots available")
+	}
+
+	pos := segment.EncodeUvarintAscending(nil, to.epoch)
+
+	snapshot := snapshots.Bucket(pos)
+	if snapshot == nil {
+		return fmt.Errorf("Rollback: snapshot not found")
+	}
+
+	revertTo, err := s.loadSnapshot(snapshot)
+	if err != nil {
+		return fmt.Errorf("Rollback: unable to load snapshot: %v", err)
+	}
+
+	// add segments referenced by loaded index snapshot to the
+	// ineligibleForRemoval map
+	for _, segSnap := range revertTo.segment {
+		filename := zapFileName(segSnap.id)
+		s.markIneligibleForRemoval(filename)
+	}
+
+	// read-only bolt transactions to be rolled back
+	_ = tx.Rollback()
```

Comment:
should we use the defer pattern we used before? i see other return paths that do not clean up the transaction

Reply:
You're right - I did miss a few return paths. The reason I didn't use defer here is because of persistence - If unsafeBatch is unset we wait for the persister to complete and there'd be a deadlock situation with the persister waiting to grab the boltdb write lock to complete the transaction. I'll address this.